In [1]:
import pandas as pd
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import load_digits
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn.tree import DecisionTreeClassifier

In [2]:
digits = load_digits()
dir(digits)

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']

In [3]:
digits.target

array([0, 1, 2, ..., 8, 9, 8])

In [4]:
digits.target_names

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [5]:
digits.target.shape

(1797,)

In [7]:
model_params = {
    "svm": {
        "model": svm.SVC(),
        "params": {
            "C": [1, 5, 10],
            "kernel": ["rbf", "linear", "sigmoid"]
        }
    },
    "random_forest":{
        "model": RandomForestClassifier(),
        "params":{
            "n_estimators": [5, 10, 20, 50, 100]
        }
    },
    "logistic_regression":{
        "model": LogisticRegression(),
        "params":{
            "C": [5, 10, 20, 50, 100],
            "max_iter": [50, 100]
        }
    },
    "gaussianNB":{
        "model": GaussianNB(),
        "params": {
            
        }
    },
    "multinomialNB":{
        "model": MultinomialNB(),
        "params": {
            
        }
    },
    "decision_tree": {
        "model": DecisionTreeClassifier(),
        "params": {
            
        }
    }
}

In [8]:
model_params.items()

dict_items([('svm', {'model': SVC(), 'params': {'C': [1, 5, 10], 'kernel': ['rbf', 'linear', 'sigmoid']}}), ('random_forest', {'model': RandomForestClassifier(), 'params': {'n_estimators': [5, 10, 20, 50, 100]}}), ('logistic_regression', {'model': LogisticRegression(), 'params': {'C': [5, 10, 20, 50, 100], 'max_iter': [50, 100]}}), ('gaussianNB', {'model': GaussianNB(), 'params': {}}), ('multinomialNB', {'model': MultinomialNB(), 'params': {}}), ('decision_tree', {'model': DecisionTreeClassifier(), 'params': {}})])

In [9]:
scores = []
for model_name, mp in model_params.items():
    clf = GridSearchCV(mp["model"], mp["params"], cv = 5, return_train_score = False)
    clf.fit(digits.data, digits.target)
    scores.append({
        "model": model_name,
        "best_score": clf.best_score_,
        "best_params": clf.best_params_
    })
result = pd.DataFrame(scores, columns = ["model", "best_score", "best_params"])
result

C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i 

C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\ACER\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i 

model  best_score                best_params
0                  svm    0.973850  {'C': 5, 'kernel': 'rbf'}
1        random_forest    0.934907      {'n_estimators': 100}
2  logistic_regression    0.917662   {'C': 5, 'max_iter': 50}
3           gaussianNB    0.806928                         {}
4        multinomialNB    0.870350                         {}
5        decision_tree    0.773542                         {}

In [10]:
result

model  best_score                best_params
0                  svm    0.973850  {'C': 5, 'kernel': 'rbf'}
1        random_forest    0.934907      {'n_estimators': 100}
2  logistic_regression    0.917662   {'C': 5, 'max_iter': 50}
3           gaussianNB    0.806928                         {}
4        multinomialNB    0.870350                         {}
5        decision_tree    0.773542                         {}

The best model is <b> SVM </b> with the best score of 97.38%. It was achieved with the parameters: {C = 5, kernel = rbf}